### PR TITLE
[exporter/awsemf] Add AppSignals metadata flag into user-agent in EMF exporter

### PR DESCRIPTION
**Pull request by jj22ee**
**Description:** <Describe what has changed.>

Cherry-picking from downstream: https://github.com/amazon-contributing/opentelemetry-collector-contrib/pull/105

This PR adds AppSignals metadata flag in user-agent in EMF exporter, so CWLogs backend can indicate EMF logs request type for the corresponding actions

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>
Unit Testing

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Review by bryan-aguilar — DISMISSED** [review]

File: internal/aws/cwlogs/cwlog_client.go
Comment (bryan-aguilar) [inline internal/aws/cwlogs/cwlog_client.go:51]: Could you make this more generalizable? Rather than only control adding a string for AppSignals it could just add any string passed into it to the user agent. The client doesn't need to know about AppSignals or not, just that it needs to add an extra string to the user agent.

File: internal/aws/cwlogs/cwlog_client.go
Comment (bryan-aguilar) [inline internal/aws/cwlogs/cwlog_client.go:41]: I think this should be `ClientOptions` rather than UserAgentOptions. It will allow us to add more options as necessary that the client may need to change.

**Review by bryan-aguilar — COMMENTED** [review]
Sorry for dismissing my review after approval. The idea of making the options more generalized came to me a bit later but I think it could be a good way to avoid a tight scope.

File: internal/aws/cwlogs/cwlog_client.go
Comment (bryan-aguilar) [inline internal/aws/cwlogs/cwlog_client.go:45]:
```suggestion
type cwLogClient struct {
	userAgentPostFix []string
}
```

I don't think this needs to be exported and it could instead be a list of strings that gets appended to the user agent header. Name could be better but I'm out of ideas atm.

File: internal/aws/cwlogs/cwlog_client.go
Comment (bryan-aguilar) [reply internal/aws/cwlogs/cwlog_client.go:51]: We may be left with a bit of tech debt if we do this though with the container insights logic in the user agent builder. I think that's okay™ for now though. Realistically a user could provide any number of extra strings to append to the user agent.

File: internal/aws/cwlogs/cwlog_client.go
Comment (bryan-aguilar) [inline internal/aws/cwlogs/cwlog_client.go:210]: With the suggestions I have made previously I don't think it has to be one or the other.

**Comment by jj22ee** [comment]
In latest commit:
- Replaced `WithEnabledAppSignals` with `WithUserAgentExtras(ex1, ex2, ....)`
  - strings provided to `WithUserAgentExtras` will be added to user agent of the cliennt
- Logic to add "AppSignals" string is now done outside of cwlog_client, now inside emf_exporter


Before this change, cwlog_client will control what strings can get added to the user agent based on flags set by the "WithOption" inputs. Now, there is a "WithOption" input to explicitly add strings to user agent. Pinging @bryan-aguilar

File: internal/aws/cwlogs/cwlog_client.go
Comment (jj22ee) [reply internal/aws/cwlogs/cwlog_client.go]: Addressed.

File: internal/aws/cwlogs/cwlog_client.go
Comment (jj22ee) [reply internal/aws/cwlogs/cwlog_client.go]: Sounds good, no need in previous logic to have a "WithOption" for every single additional agent string.
With your suggestion, I replaced this with: `WithUserAgentExtras(userAgentExtras ...string`)

File: internal/aws/cwlogs/cwlog_client.go
Comment (jj22ee) [reply internal/aws/cwlogs/cwlog_client.go]: Updated using your suggestion with names: `cwLogClientConfig` and `userAgentExtras`

File: exporter/awsemfexporter/config.go
Comment (Aneurysm9) [inline exporter/awsemfexporter/config.go:141]: This shouldn't be exported.

Comment (jj22ee) [reply exporter/awsemfexporter/config.go]: Removed export of this function.

**Review by Aneurysm9 — APPROVED** [review]

**Review by mxiamxia — APPROVED** [review]

**Comment by jj22ee** [comment]
Thanks! Can this PR be marked as ready for merge?

**Comment by jj22ee** [comment]
pinging @jpkrohling @bogdandrutu @bryan-aguilar for merge

**Comment by jj22ee** [comment]
We will need a different approver/maintainer to help approve/merge this PR, afaik @bryan-aguilar may not be available to do this as of now.

Can you take a look @jpkrohling @bogdandrutu

**Comment by jj22ee** [comment]
@fatsheep9146 Could you help take a look for approving our PR for merge? We have sufficient reviews from AWS side, and @bryan-aguilar is currently not be available to help continue his review afaik.

**Review by dmitryax — APPROVED** [review]

**Pull request merged**
